Source/Core/Common/Config/Config.cpp

```diff
@@ -5,51 +5,61 @@
 #include <algorithm>
 #include <list>
 #include <map>
+#include <shared_mutex>
 
 #include "Common/Config/Config.h"
 
 namespace Config
 {
+using Layers = std::map<LayerType, std::shared_ptr<Layer>>;
+
+static std::shared_mutex s_layersRWLock;
 static Layers s_layers;
 static std::list<ConfigChangedCallback> s_callbacks;
 static u32 s_callback_guards = 0;
 
-Layers* GetLayers()
-{
-  return &s_layers;
-}
-
-void AddLayer(std::unique_ptr<Layer> layer)
+void AddLayerInternal(std::shared_ptr<Layer> layer)
 {
-  s_layers[layer->GetLayer()] = std::move(layer);
+  const Config::LayerType layerType = layer->GetLayer();
+  {
+    std::unique_lock lock(s_layersRWLock);
+    s_layers[layerType] = std::move(layer);
+  }
   InvokeConfigChangedCallbacks();
 }
 
 void AddLayer(std::unique_ptr<ConfigLayerLoader> loader)
 {
-  AddLayer(std::make_unique<Layer>(std::move(loader)));
+  AddLayerInternal(std::make_shared<Layer>(std::move(loader)));
 }
 
-Layer* GetLayer(LayerType layer)
+std::shared_ptr<Layer> GetLayer(LayerType layer)
 {
-  if (!LayerExists(layer))
-    return nullptr;
-  return s_layers[layer].get();
+  std::shared_ptr<Layer> result;
```

Comment:
This can be moved closer to its assignment.

Reply:
I deliberately placed it here to construct it prior to locking the mutex.

Reply:
For what purpose? It's a non-conditional assignment to nullptr.